delivery_postlogistics/models/stock_picking.py

```diff
@@ -0,0 +1,286 @@
+# Copyright 2013-2016 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+import base64
+from operator import attrgetter
+
+from odoo import _, api, exceptions, fields, models
+
+from ..postlogistics.web_service import PostlogisticsWebService
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    delivery_type = fields.Selection(related="carrier_id.delivery_type")
+    # TODO: Change the name of this field
+    postlogistics_option_ids = fields.Many2many(
+        comodel_name="postlogistics.delivery.carrier.option", string="Options"
+    )
+
+    delivery_fixed_date = fields.Date(
+        "Fixed delivery date", help="Specific delivery date (ZAW3217)"
+    )
+    delivery_place = fields.Char(
+        "Delivery Place", help="For Deposit item service (ZAW3219)"
+    )
+    delivery_phone = fields.Char(
+        "Phone", help="For notify delivery by telephone (ZAW3213)"
+    )
+    delivery_mobile = fields.Char(
+        "Mobile", help="For notify delivery by telephone (ZAW3213)"
+    )
+
+    @api.onchange("carrier_id")
+    def onchange_carrier_id(self):
+        """ Inherit this method in your module """
+        if not self.carrier_id:
+            return
+        # This can look useless as the field carrier_code and
+        # carrier_type are related field. But it's needed to fill
+        # this field for using this fields in the view. Indeed the
+        # module that depend of delivery base can hide some field
+        # depending of the type or the code
+        carrier = self.carrier_id
+        self.update({"delivery_type": carrier.delivery_type})
+        default_options = carrier.default_options()
+        self.postlogistics_option_ids = default_options
+        result = {
+            "domain": {
+                "postlogistics_option_ids": [
+                    ("id", "in", carrier.postlogistics_available_option_ids.ids)
+                ]
+            }
+        }
+        return result
+
+    @api.onchange("postlogistics_option_ids")
+    def onchange_postlogistics_option_ids(self):
+        if not self.carrier_id:
+            return
+        carrier = self.carrier_id
+        for available_option in carrier.postlogistics_available_option_ids:
+            if (
+                available_option.mandatory
+                and available_option not in self.postlogistics_option_ids
+            ):
+                # XXX the client does not allow to modify the field that
+                # triggered the onchange:
+                # https://github.com/odoo/odoo/issues/2693#issuecomment-56825399
+                # Ideally we should add the missing option
+                raise exceptions.UserError(
```

Comment:
This prevents to change the carrier as this trigger this onchange. If postlogictics has a mandatory option, then this error will raise.

Reply:
Can you dig this issue. Configure postlogistics with a mandatory option. Take a picking without carrier and set the postlogistic carrier. This triggers the onchange on carrier that will fill the mandatory option as defined here https://github.com/OCA/delivery-carrier/pull/323/files#diff-7d5856a39a59766b6cd8d580e267bd8b19f48ff5ceccb17d7243e90632feef21R46
then this will trigger this onchange but the error should not be raised as the option as already been added.

Reply:
The error will also raise, when a new SO is created with this carrier or if the missing mandatory option is tried to be add to an existing SO with this carrier.

Reply:
the field `postlogistics_option_ids` on picking will be removed soon, so we will remove this method also

Reply:
fields postlogistics_option_ids and the function related to this field removed in the latest update